capsules/src/touch.rs

```diff
@@ -0,0 +1,326 @@
+            // | Touch 0                                                                             | Touch 1  ...
+            2 => {
+                if self.multi_touch.is_some() {
+                    self.apps
+                        .enter(appid, |app, _| {
+                            app.events_buffer = slice;
+                            ReturnCode::SUCCESS
+                        })
+                        .unwrap_or_else(|err| err.into())
+                } else {
+                    ReturnCode::ENOSUPPORT
+                }
+            }
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    fn subscribe(
+        &self,
+        subscribe_num: usize,
+        callback: Option<Callback>,
+        app_id: AppId,
+    ) -> ReturnCode {
+        match subscribe_num {
+            // subscribe to touch
+            0 => self
+                .apps
+                .enter(app_id, |app, _| {
+                    app.touch_callback = callback;
+                    ReturnCode::SUCCESS
+                })
+                .unwrap_or_else(|err| err.into()),
+
+            // subscribe to gestures
+            1 => self
+                .apps
+                .enter(app_id, |app, _| {
+                    app.gesture_callback = callback;
+                    ReturnCode::SUCCESS
+                })
+                .unwrap_or_else(|err| err.into()),
+
+            // subscribe to multi touch
+            2 => {
+                if self.multi_touch.is_some() {
+                    self.apps
+                        .enter(app_id, |app, _| {
+                            app.multi_touch_callback = callback;
+                            ReturnCode::SUCCESS
+                        })
+                        .unwrap_or_else(|err| err.into())
+                } else {
+                    ReturnCode::ENOSUPPORT
+                }
+            }
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    fn command(
+        &self,
+        command_num: usize,
+        _data1: usize,
+        _data2: usize,
+        _appid: AppId,
+    ) -> ReturnCode {
+        match command_num {
+            0 =>
+            // This driver exists.
+            {
+                ReturnCode::SUCCESS
+            }
+
+            // Touch Enable
+            1 => {
+                if let Some(touch) = self.touch {
+                    touch.enable()
+                } else if let Some(multi_touch) = self.multi_touch {
+                    multi_touch.enable()
+                } else {
+                    ReturnCode::ENODEVICE
+                }
+            }
+            // Touch Disable
+            2 => {
+                if let Some(touch) = self.touch {
+                    touch.disable()
+                } else if let Some(multi_touch) = self.multi_touch {
+                    multi_touch.disable()
+                } else {
```

Comment:
You have a point here, this was for allowing apps to really disable it. What about allowing this if there are no subscriptions?

Reply:
I think there should be a way to do that, but it shouldn't return an error if it can't disable anything. That would tell the app that another app is using the touchscreen, and leak that information. I think it would overall be easier to just disable the screen automatically if nothing is using it, and not have this API.

Reply:
It is now enabled and disabled automatically